File: sample/rs/15_wave_interference_loop.rs
use crate::math;
use crate::time::perf_counter;
use crate::pytra::utils::gif::grayscale_palette;
use crate::pytra::utils::gif::save_gif;

// 15: Sample that renders wave interference animation and writes a GIF.

fn run_15_wave_interference_loop() {
    let w = 320;
    let h = 240;
    let frames_n = 96;
    let out_path = "sample/out/15_wave_interference_loop.gif";
    
    let start = perf_counter();
    let mut frames: Vec<Vec<u8>> = vec![];
    
    let mut t: i64 = 0;
    while t < frames_n {
        let mut frame = bytearray((w * h));
        let phase = (t * 0.12);
        let mut y: i64 = 0;
        while y < h {
            let row_base = (y * w);
            let mut x: i64 = 0;
            while x < w {
                let dx = (x - 160);
                let dy = (y - 120);
                let v = (((math.sin((((x + (t * 1.5))) * 0.045)) + math.sin((((y - (t * 1.2))) * 0.04))) + math.sin(((((x + y)) * 0.02) + phase))) + math.sin(((math.sqrt(((dx * dx) + (dy * dy))) * 0.08) - (phase * 1.3))));
                let mut c = (((v + 4.0)) * ((255.0 / 8.0))) as i64;
                if c < 0 {
                    c = 0;
                }
                if c > 255 {
                    c = 255;
                }
                frame[(row_base + x) as usize] = c;
                x += 1;
            }
            y += 1;
        }
        frames.push(bytes(frame));
        t += 1;
    }
    save_gif(out_path, w, h, frames, grayscale_palette());
    let elapsed = (perf_counter() - start);
    println!("{:?}", ("output:", out_path));
    println!("{:?}", ("frames:", frames_n));
    println!("{:?}", ("elapsed_sec:", elapsed));
}

fn main() {
    run_15_wave_interference_loop();
}

File: sample/cpp/15_wave_interference_loop.cpp
#include "runtime/cpp/native/core/py_runtime.h"
#include "runtime/cpp/native/core/process_runtime.h"
#include "runtime/cpp/native/core/scope_exit.h"

#include "generated/built_in/io_ops.h"
#include "generated/std/math.h"
#include "generated/std/time.h"
#include "generated/utils/gif.h"

// 15: Sample that renders wave interference animation and writes a GIF.

void run_15_wave_interference_loop() {
    int64 w = 320;
    int64 h = 240;
    int64 frames_n = 96;
    str out_path = "sample/out/15_wave_interference_loop.gif";
    
    float64 start = pytra::std::time::perf_counter();
    rc<list<bytes>> frames = rc_list_from_value(list<bytes>{});
    
    for (int64 t = 0; t < frames_n; ++t) {
        bytearray frame = bytearray(w * h);
        float64 phase = static_cast<float64>(t) * 0.12;
        for (int64 y = 0; y < h; ++y) {
            int64 row_base = y * w;
            for (int64 x = 0; x < w; ++x) {
                int64 dx = x - 160;
                int64 dy = y - 120;
                float64 v = pytra::std::math::sin((static_cast<float64>(x) + static_cast<float64>(t) * 1.5) * 0.045) + pytra::std::math::sin((static_cast<float64>(y) - static_cast<float64>(t) * 1.2) * 0.04) + pytra::std::math::sin((static_cast<float64>(x + y)) * 0.02 + phase) + pytra::std::math::sin(pytra::std::math::sqrt(dx * dx + dy * dy) * 0.08 - phase * 1.3);
                int64 c = static_cast<int64>((v + 4.0) * (255.0 / 8.0));
                if (c < 0)
                    c = 0;
                if (c > 255)
                    c = 255;
                frame[row_base + x] = c;
            }
        }
        rc_list_ref(frames).append(frame);
    }
    pytra::utils::gif::save_gif(out_path, w, h, rc_list_ref(frames), pytra::utils::gif::grayscale_palette(), 4, 0);
    float64 elapsed = pytra::std::time::perf_counter() - start;
    py_print("output:", out_path);
    py_print("frames:", frames_n);
    py_print("elapsed_sec:", elapsed);
}

int main(int argc, char** argv) {
    pytra_configure_from_argv(argc, argv);
    run_15_wave_interference_loop();
    return 0;
}
